session.rst

```diff
@@ -1,15 +1,276 @@
+   single: HTTP
+   single: HttpFoundation, Sessions
+
 Sessions
 ========
 
-Symfony provides a session object and several utilities that you can use to
-store information about the user between requests.
+The Symfony HttpFoundation component has a very powerful and flexible session
+subsystem which is designed to provide session management that you can use to
+store information about the user between requests through a clear
+object-oriented interface using a variety of session storage drivers.
+
+Symfony sessions are designed to replace the usage of the ``$_SESSION`` super
+global and native PHP functions related to manipulating the session like
+``session_start()``, ``session_regenerate_id()``, ``session_id()``,
+``session_name()``, and ``session_destroy()``.
+
+.. note::
+
+    Sessions are only started if you read or write from it.
+
+Installation
+------------
+
+You need to install the HttpFoundation component to handle sessions:
+
+.. code-block:: terminal
+
+    $ composer require symfony/http-foundation
+
+.. _session-intro:
+
+Basic Usage
+-----------
+
+The session is available through the ``Request`` object and the ``RequestStack``
+service. Symfony injects the ``request_stack`` service in services and controllers
+if you type-hint an argument with :class:`Symfony\\Component\\HttpFoundation\\RequestStack`::
+
+.. configuration-block::
+
+    .. code-block:: php-symfony
+
+        use Symfony\Component\HttpFoundation\RequestStack;
+
+        class SomeService
+        {
+            private $requestStack;
+
+            public function __construct(RequestStack $requestStack)
+            {
+                $this->requestStack = $requestStack;
+
+                // Accessing the session in the constructor is *NOT* recommended, since
+                // it might not be accessible yet or lead to unwanted side-effects
+                // $this->session = $requestStack->getSession();
+            }
+
+            public function someMethod()
+            {
+                $session = $this->requestStack->getSession();
+
+                // ...
+            }
+        }
+
+    .. code-block:: php-standalone
+
+        use Symfony\Component\HttpFoundation\Session\Attribute\AttributeBag;
+        use Symfony\Component\HttpFoundation\Session\Session;
+        use Symfony\Component\HttpFoundation\Session\Storage\NativeSessionStorage;
+
+        $session = new Session();
+        $session->start();
+
+        // Pass a storage and a bag implementation explicitely
+        $session = new Session(new NativeSessionStorage(), new AttributeBag());
+
+From a Symfony controller, type-hint an argument with
+:class:`Symfony\\Component\\HttpFoundation\\RequestStack` or
+:class:`Symfony\\Component\\HttpFoundation\\Request`::
+
+    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HttpFoundation\RequestStack;
+    use Symfony\Component\HttpFoundation\Response;
+
+    public function index(RequestStack $requestStack): Response
```

Comment:
Showing two different ways to access the session in the same code snippet might be confusing to, as it is not clear why there are two ways and when to use which.

Maybe we can add a note that Controllers commonly get the session from the `Request` whereas services might want to use the `RequestStack` instead, so that you don't have to pass around the request and/or session between services? Alternatively, it might make sense to update the examples to have one controller- and one service-example showing both ways independently.

Reply:
Done. I have remove the RequestStack example for the controller as this is never something you need.
As we already have an example for a service just above, I think it's good now.